code/datums/components/trapdoor.dm

```diff
@@ -0,0 +1,171 @@
+	UnregisterSignal(parent, COMSIG_TURF_CHANGE)
+
+///called by linking remotes to tie an assembly to the trapdoor
+/datum/component/trapdoor/proc/link_with_assembly(obj/item/assembly/trapdoor/assembly)
+	src.assembly = assembly
+	GLOB.unlinked_trapdoors -= src
+	assembly.linked = TRUE
+	RegisterSignal(assembly, COMSIG_ASSEMBLY_PULSED, .proc/toggle_trapdoor)
+
+///signal called by our assembly being pulsed
+/datum/component/trapdoor/proc/toggle_trapdoor(datum/source)
+	SIGNAL_HANDLER
+	if(!IS_OPEN(parent))
+		try_opening()
+	else
+		try_closing()
+
+///signal called by turf changing
+/datum/component/trapdoor/proc/turf_changed_pre(datum/source, path, new_baseturfs, flags, post_change_callbacks)
+	SIGNAL_HANDLER
+	var/turf/open/dying_trapdoor = parent
+	if(!IS_OPEN(dying_trapdoor) && !IS_OPEN(path)) //not a process of the trapdoor, so this trapdoor has been destroyed
+		dying_trapdoor.visible_message("<span class='warning'>the trapdoor mechanism in [dying_trapdoor] is broken!</span>")
+		assembly.forceMove(dying_trapdoor) //drops components that went into this
+		return
+	post_change_callbacks += CALLBACK(src, .proc/turf_changed_post, assembly, trapdoor_turf_path)
+
+/**
+ * # turf_changed_post
+ *
+ * wrapper that applies the trapdoor to the new turf (created by the last trapdoor)
+ * change da turf my final callback. Goodbye
+ */
+/datum/component/trapdoor/proc/turf_changed_post(turf/new_turf, assembly, trapdoor_turf_path)
+	new_turf.AddComponent(/datum/component/trapdoor, starts_open = TRUE, assembly, trapdoor_turf_path)
+
+/**
+ * # try_opening
+ *
+ * small proc for opening the turf into openspace
+ * there are no checks for opening a trapdoor, but closed has some
+ */
+/datum/component/trapdoor/proc/try_opening()
+	var/turf/open/trapdoor_turf = parent
+	trapdoor_turf.visible_message("<span class='warning'>[trapdoor_turf] swings open!</span>")
+	trapdoor_turf.ChangeTurf(/turf/open/openspace, flags = CHANGETURF_INHERIT_AIR)
+
+/**
+ * # try_closing
+ *
+ * small proc for closing the turf back into what it should be
+ * trapdoor can be blocked by building things on the openspace turf
+ */
+/datum/component/trapdoor/proc/try_closing()
+	var/turf/open/trapdoor_turf = parent
+	var/obj/structure/lattice/blocking = locate() in trapdoor_turf.contents
+	if(blocking)
+		trapdoor_turf.visible_message("<span class='warning'>The trapdoor mechanism in [trapdoor_turf] tries to shut, but is jammed by [blocking]!</span>")
+		return
+	trapdoor_turf.visible_message("<span class='warning'>The trapdoor mechanism in [trapdoor_turf] swings shut!</span>")
+	trapdoor_turf.visible_message("<span class='warning'>[trapdoor_turf] swings open!</span>")
+	trapdoor_turf.ChangeTurf(trapdoor_turf_path, flags = CHANGETURF_INHERIT_AIR)
+
+#undef IS_OPEN
+
+/obj/item/assembly/trapdoor
+	name = "trapdoor controller"
+	desc = "A sinister-looking controller for a trapdoor."
+	var/linked = FALSE
+
+#define TRAPDOOR_LINKING_SEARCH_RANGE 7
+
+/obj/item/trapdoor_remote
```

Comment:
needs a subtype with internals already installed, this is basically what a crafted remote would be (requiring you to fill the internals before it can be used

Reply:
which reminds me, this needs an attackby to insert internals and an altclick to forget the trapdoor it is linked to.